bluemira/builders/tf_coils.py

```diff
@@ -140,17 +143,397 @@ def _make_single_circuit(self, wire):
+        self.points = Coordinates(
+            [
+                coords.points[arg_x_min],
+                coords.points[arg_x_max],
+                coords.points[arg_z_min],
+                coords.points[arg_z_max],
+            ]
+        )
+
+
+class FixedSelector(RipplePointSelector):
+    """
+    Specified points at which to constrain the ripple, overrides any information
+    relating directly to the separatrix.
+
+    Parameters
+    ----------
+    points:
+        Points at which the ripple should be constrained.
+    """
+
+    def __init__(self, points: Coordinates):
+        self.points = points
+
+
+class MaximiseSelector(RipplePointSelector):
+    """
+    Finds and constrains the maximum ripple along the specified wire during
+    each minimisation function call.
+
+    Parameters
+    ----------
+    wire:
+        Wire along which to constrain the ripple
+    """
+
+    def __init__(self):
+        self.points = None
+
+    def set_wire(self, wire):
+        """
+        Set the wire along which the points will be selected
+
+        Parameters
+        ----------
+        wire
+            Wire along which the points will be selected
+        """
+        super().set_wire(wire)
+        points = wire.discretize(byedges=True, ndiscr=200)
+        arg_x_max = np.argmax(points.x)
+        x_max_point = points[:, arg_x_max]
+        self._alpha_0 = wire.parameter_at(x_max_point, tolerance=10 * EPS)
+
+    def make_ripple_constraint(
+        self, parameterisation, solver, TF_ripple_limit, rip_con_tol
+    ) -> OptimisationConstraint:
+        """
+        Make the ripple OptimisationConstraint
+        """
+        return OptimisationConstraint(
+            self.constrain_max_ripple,
+            f_constraint_args={
+                "parameterisation": parameterisation,
+                "solver": solver,
+                "TF_ripple_limit": TF_ripple_limit,
+                "lcfs_wire": self.wire,
+                "alpha_0": self._alpha_0,
+                # I'm sorry...
+                "this": self,
+            },
+            tolerance=rip_con_tol * np.ones(2),
+        )
+
+    @staticmethod
+    def constrain_max_ripple(
+        constraint,
+        vector,
+        grad,
+        parameterisation,
+        solver,
+        lcfs_wire: BluemiraWire,
+        alpha_0: float,
+        TF_ripple_limit: float,
+        this,
+        ad_args=None,
+    ):
+        """
+        Ripple constraint function
+
+        Parameters
+        ----------
+        constraint: np.ndarray
+            Constraint vector (updated in place)
+        vector: np.ndarray
+            Variable vector
+        grad: np.ndarray
+            Jacobian matrix of the constraint (updated in place)
+        parameterisation: GeometryParameterisation
+            Geometry parameterisation
+        solver: ParameterisedHelmholtzSolver
+            TF ripple solver
+        TF_ripple_limit: float
+            Maximum allowable TF ripple
```

Comment:
Can we finish off the docstring, some of the arguments aren't documented, alternatively this method could be made private (`_constrain...`) as it feels like it wont be used outside of the class.